server-utils/src/cors.rs

```diff
@@ -201,18 +246,99 @@ pub fn get_cors_header(origin: Option<&str>, host: Option<&str>, allowed: &Optio
 						}
 					})
 					.map(|_| AccessControlAllowOrigin::Value(Origin::parse(origin)))
-					.map(CorsHeader::Ok).unwrap_or(CorsHeader::Invalid)
+					.map(AllowOrigin::Ok).unwrap_or(AllowOrigin::Invalid)
+				},
+			}
+		},
+	}
+}
+
+/// Validates if the `AccessControlAllowedHeaders` in the request are allowed.
+pub fn get_cors_allow_headers(request_headers: &Headers, cors_allow_headers: &AccessControlAllowHeaders) -> AllowHeaders {
+	// Check if the header fields which were sent in the request are required
+	if let AccessControlAllowHeaders::Only(only) = cors_allow_headers {
+		let are_all_allowed = request_headers.iter()
+			.all(|header| {
+				only.contains(&Ascii::new(header.name().to_owned())) ||
+				get_always_allowed_headers().contains(&Ascii::new(header.name()))
+			});
+
+		if !are_all_allowed {
+			return AllowHeaders::Invalid;
+		}
+	}
+
+	// Check if `AccessControlRequestHeaders` contains fields which were allowed
+	match request_headers.get::<AccessControlRequestHeaders>() {
+		None => {
+			// No fields were requested, no comparison necessary
+			match cors_allow_headers {
+				AccessControlAllowHeaders::Any => AllowHeaders::NotRequired,
+
+				// The fields which can be used are constrainted, but since it
+				// was asked for none we don't return any.
+				AccessControlAllowHeaders::Only(_) => {
+					let empty = header::AccessControlAllowHeaders(vec![]);
+					AllowHeaders::Ok(empty)
 				},
 			}
 		},
+		Some(requested) => {
+			// "requested" contains the fields for which it is inquired to know
+			// if they can be used.
+
+			let echo = AllowHeaders::Ok(
+				header::AccessControlAllowHeaders(requested.deref().clone())
+			);
+
+			match cors_allow_headers {
+				AccessControlAllowHeaders::Any => {
+					// Any field is allowed. Our response are the fields about which
+					// the request inquired.
+					echo
+				},
+
+				AccessControlAllowHeaders::Only(only) => {
+					let are_all_allowed = requested.iter()
+						.all(|header| {
+							let name = &Ascii::new(header.as_ref());
+							only.contains(header) || get_always_allowed_headers().contains(name)
+						});
+
+					if !are_all_allowed {
+						return AllowHeaders::Invalid;
+					}
+
+					echo
+				}
+			}
+		}
 	}
 }
 
+/// Returns headers which are always allowed.
+pub fn get_always_allowed_headers() -> Vec<Ascii<&'static str>> {
```

Comment:
I think it would be better to do a `HashSet` here instead of doing linear search every time. The list of headers looks good.